File: 55_Life/javascript/life.js
// LIFE
//
// Converted from BASIC to Javascript by Oscar Toledo G. (nanochess)
//

function print(str)
{
    document.getElementById("output").appendChild(document.createTextNode(str));
}

function input()
{
    var input_element;
    var input_str;
    
    return new Promise(function (resolve) {
                       input_element = document.createElement("INPUT");
                       
                       print("? ");
                       input_element.setAttribute("type", "text");
                       input_element.setAttribute("length", "50");
                       document.getElementById("output").appendChild(input_element);
                       input_element.focus();
                       input_str = undefined;
                       input_element.addEventListener("keydown", function (event) {
                                                      if (event.keyCode == 13) {
                                                      input_str = input_element.value;
                                                      document.getElementById("output").removeChild(input_element);
                                                      print(input_str);
                                                      print("\n");
                                                      resolve(input_str);
                                                      }
                                                      });
                       });
}

function tab(space)
{
    var str = "";
    while (space-- > 0)
        str += " ";
    return str;
}

var bs = [];
var a = [];

// Main program
async function main()
{
    print(tab(34) + "LIFE\n");
    print(tab(15) + "CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY\n");
    print("\n");
    print("\n");
    print("\n");
    print("ENTER YOUR PATTERN:\n");
    x1 = 1;
    y1 = 1;
    x2 = 24;
    y2 = 70;
    for (c = 1; c <= 24; c++) {
        bs[c] = "";
        a[c] = [];
        for (d = 1; d <= 70; d++)
            a[c][d] = 0;
    }
    c = 1;
    while (1) {
        bs[c] = await input();
        if (bs[c] == "DONE") {
            bs[c] = "";
            break;
        }
        if (bs[c].substr(0, 1) == ".")
            bs[c] = " " + bs[c].substr(1);
        c++;
    }
    c--;
    l = 0;
    for (x = 1; x <= c - 1; x++) {
        if (bs[x].length > l)
            l = bs[x].length;
    }
    x1 = 11 - (c >> 1);
    y1 = 33 - (l >> 1);
    p = 0;
    for (x = 1; x <= c; x++) {
        for (y = 1; y <= bs[x].length; y++) {
            if (bs[x][y - 1] != " ") {
                a[x1 + x][y1 + y] = 1;
                p++;
            }
        }
    }
    print("\n");
    print("\n");
    print("\n");
    i9 = false;
    g = 0;
    while (g < 100) {
        print("GENERATION: " + g + " POPULATION: " + p + " ");
        if (i9)
            print("INVALID!");
        x3 = 24;
        y3 = 70;
        x4 = 1;
        y4 = 1;
        p = 0;
        g++;
        for (x = 1; x <= x1 - 1; x++)
            print("\n");
        for (x = x1; x <= x2; x++) {
            print("\n");
            str = "";
            for (y = y1; y <= y2; y++) {
                if (a[x][y] == 2) {
                    a[x][y] = 0;
                    continue;
                } else if (a[x][y] == 3) {
                    a[x][y] = 1;
                } else if (a[x][y] != 1) {
                    continue;
                }
                while (str.length < y)
                    str += " ";
                str += "*";
                if (x < x3)
                    x3 = x;
                if (x > x4)
                    x4 = x;
                if (y < y3)
                    y3 = y;
                if (y > y4)
                    y4 = y;
            }
            print(str);
        }
        for (x = x2 + 1; x <= 24; x++)
            print("\n");
        x1 = x3;
        x2 = x4;
        y1 = y3;
        y2 = y4;
        if (x1 < 3) {
            x1 = 3;
            i9 = true;
        }
        if (x2 > 22) {
            x2 = 22;
            i9 = true;
        }
        if (y1 < 3) {
            y1 = 3;
            i9 = true;
        }
        if (y2 > 68) {
            y2 = 68;
            i9 = true;
        }
        p = 0;
        for (x = x1 - 1; x <= x2 + 1; x++) {
            for (y = y1 - 1; y <= y2 + 1; y++) {
                c = 0;
                for (i = x - 1; i <= x + 1; i++) {
                    for (j = y - 1; j <= y + 1; j++) {
                        if (a[i][j] == 1 || a[i][j] == 2)
                            c++;
                    }
                }
                if (a[x][y] == 0) {
                    if (c == 3) {
                        a[x][y] = 3;
                        p++;
                    }
                } else {
                    if (c < 3 || c > 4) {
                        a[x][y] = 2;
                    } else {
                        p++;
                    }
                }
            }
        }
        x1--;
        y1--;
        x2++;
        y2++;
    }
}

main();

File: 55_Life/java/src/java/Life.java
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

/**
 * The Game of Life class.<br>
 * <br>
 * Mimics the behaviour of the BASIC version, however the Java code does not have much in common with the original.
 * <br>
 * Differences in behaviour:
 * <ul>
 *     <li>Input supports the "." character, but it's optional.</li>
 *     <li>Input regarding the "DONE" string is case insensitive.</li>
 * </ul>
 */
public class Life {

    private static final byte DEAD  = 0;
    private static final byte ALIVE = 1;
    private static final String NEWLINE = "\n";

    private final Scanner consoleReader = new Scanner(System.in);

    private final byte[][] matrix = new byte[21][67];
    private int generation = 0;
    private int population = 0;
    boolean stopAfterGen = false;
    boolean invalid = false;


    public Life(String[] args) {
        parse(args);
    }

    private void parse(String[] args) {
        for (String arg : args) {
            if ("-s".equals(arg)) {
                stopAfterGen = true;
                break;
            }
        }
    }

    private void start() {
        printGameHeader();
        readPattern();
        while (true) {
            printGeneration();
            advanceToNextGeneration();
            if (stopAfterGen) {
                consoleReader.nextLine();
            }
        }
    }

    private void advanceToNextGeneration() {
        // store all cell transitions in a list, i.e. if a dead cell becomes alive, or a living cell dies
        List<Transition> transitions = new ArrayList<>();
        // there's still room for optimization: instead of iterating over all cells in the matrix,
        // we could consider only the section containing the pattern(s), as in the BASIC version
        for (int y = 0; y < matrix.length; y++) {
            for (int x = 0; x < matrix[y].length; x++) {
                int neighbours = countNeighbours(y, x);
                if (matrix[y][x] == ALIVE) {
                    if (neighbours < 2 || neighbours > 3) {
                        transitions.add(new Transition(y, x, DEAD));
                        population--;
                    }
                } else { // cell is dead
                    if (neighbours == 3) {
                        if (x < 2 || x > 67 || y < 2 || y > 21) {
                            invalid = true;
                        }
                        transitions.add(new Transition(y, x, ALIVE));
                        population++;
                    }
                }
            }
        }
        // apply all transitions to the matrix
        transitions.forEach(t -> matrix[t.y()][t.x()] = t.newState());
        generation++;
    }

    private int countNeighbours(int y, int x) {
        int neighbours = 0;
        for (int row = Math.max(y - 1, 0); row <= Math.min(y + 1, matrix.length - 1); row++) {
            for (int col = Math.max(x - 1, 0); col <= Math.min(x + 1, matrix[row].length - 1); col++) {
                if (row == y && col == x) {
                    continue;
                }
                if (matrix[row][col] == ALIVE) {
                    neighbours++;
                }
            }
        }
        return neighbours;
    }

    private void readPattern() {
        System.out.println("ENTER YOUR PATTERN:");
        List<String> lines = new ArrayList<>();
        String line;
        int maxLineLength = 0;
        boolean reading = true;
        while (reading) {
            System.out.print("? ");
            line = consoleReader.nextLine();
            if (line.equalsIgnoreCase("done")) {
                reading = false;
            } else {
                // optional support for the '.' that is needed in the BASIC version
                lines.add(line.replace('.', ' '));
                maxLineLength = Math.max(maxLineLength, line.length());
            }
        }
        fillMatrix(lines, maxLineLength);
    }

    private void fillMatrix(List<String> lines, int maxLineLength) {
        float xMin = 33 - maxLineLength / 2f;
        float yMin = 11 - lines.size() / 2f;
        for (int y = 0; y < lines.size(); y++) {
            String line = lines.get(y);
            for (int x = 1; x <= line.length(); x++) {
                if (line.charAt(x-1) == '*') {
                    matrix[floor(yMin + y)][floor(xMin + x)] = ALIVE;
                    population++;
                }
            }
        }
    }

    private int floor(float f) {
        return (int) Math.floor(f);
    }

    private void printGameHeader() {
        printIndented(34, "LIFE");
        printIndented(15, "CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY");
        System.out.println(NEWLINE.repeat(3));
    }

    private void printIndented(int spaces, String str) {
        System.out.println(" ".repeat(spaces) + str);
    }

    private void printGeneration() {
        printGenerationHeader();
        for (int y = 0; y < matrix.length; y++) {
            for (int x = 0; x < matrix[y].length; x++) {
                System.out.print(matrix[y][x] == 1 ? "*" : " ");
            }
            System.out.println();
        }
    }

    private void printGenerationHeader() {
        String invalidText = invalid ? "INVALID!" : "";
        System.out.printf("GENERATION: %-13d POPULATION: %d %s\n", generation, population, invalidText);
    }

    /**
     * Main method that starts the program.
     *
     * @param args the command line arguments.
     * @throws Exception if something goes wrong.
     */
    public static void main(String[] args) throws Exception {
        new Life(args).start();
    }

}

/**
 * Represents a state change for a single cell within the matrix.
 *
 * @param y the y coordinate (row) of the cell
 * @param x the x coordinate (column) of the cell
 * @param newState the new state of the cell (either DEAD or ALIVE)
 */
record Transition(int y, int x, byte newState) { }
